### PR TITLE
Resource manager for bitmaps on Linux

### PR DESCRIPTION
**Pull request by jarkkojs**
Piggy pack bitmap files to a blob and link in to the plugin binary.
Implement findMemoryBitmap() to look them up in CScalableBitmap and use
IPlatformBitmap::createFromMemory() to load them.

Cleaned up the build-linux.sh from some inconsistencies while doing this
i.e.  '--install' should actually be '--install-local'. There should
eventually be a separate '--install' option used for packaging that deploys
files under /usr/share. The local install for is for development purposes.
Removed '--build-and-install' as it is just as easy to chain commands with
'&&'.

### CODE REVIEW

**Comment by jsakkine** [comment]
The prints are temporary. I will clean this up once it works. It is 2:30AM right now, so decided to stop here and put visible what I have right now.

**Comment by baconpaul** [comment]
Quick scan from my phone and looks like great progress!

**Comment by baconpaul** [comment]
Oh and one other thing on that throw

There are some bitmap ids which are addentried but not used. I think they don’t exist in the export. So catching that throw and returning null from the bitmap loader is probably the right behavior. The code above is all null sensitive

Or we can remove the add entry callsnfor the unused bitmaps.

I haven’t done a full audit of add bitmap which were never get bitmap. I just let add bitmap in missing images fail and didn’t worry by being fail sensitive above

Hope that helps

**Comment by baconpaul** [comment]
For instance

```
paul:~/dev/music/surge$ grep 110 src/common/resource.h 
#define IDB_OSCSWITCH 110
paul:~/dev/music/surge$ grep -r IDB_OSCSWITCH src/ | grep -v resource.h
src//common/gui/SurgeBitmaps.cpp:      addEntry(IDB_OSCSWITCH); 
```

compared with

```
paul:~/dev/music/surge$ grep 164 src/common/resource.h 
#define IDI_ICON1 164
#define IDI_SURGE 164
#define IDB_BUTTON_MENU 164 
paul:~/dev/music/surge$ grep -r IDB_BUTTON_MENU src | grep -v resource.h
src/common/gui/SurgeBitmaps.cpp:      addEntry(IDB_BUTTON_MENU);
src/common/gui/SurgeGUIEditor.cpp:                                             getSurgeBitmap(IDB_BUTTON_MENU), nopoint, false);
src/windows/surge.rc:     IDB_BUTTON_MENU PNG "resources/bitmaps/bmp00164.png"    
src/windows/scalableui.rc:IDB_BUTTON_MENU_SCALE_100 PNG "assets/original-vector/exported/bmp00164.png"
src/windows/scalableui.rc:IDB_BUTTON_MENU_SCALE_150 PNG "assets/original-vector/exported/bmp00164@15x.png"
src/windows/scalableui.rc:IDB_BUTTON_MENU_SCALE_200 PNG "assets/original-vector/exported/bmp00164@2x.png"
src/windows/scalableui.rc:IDB_BUTTON_MENU_SCALE_300 PNG "assets/original-vector/exported/bmp00164@3x.png"
src/windows/scalableui.rc:IDB_BUTTON_MENU_SCALE_400 PNG "assets/original-vector/exported/bmp00164@4x.png"
```

so the form for IDB_OSCSWITCH is either to fail less loudly or remove the addEntry.

I'm fine with either TBH! But failing silently means you don't have to do these checks for each one as you go.

**Comment by jsakkine** [comment]
Now it works, but I will clean this up a bit...

File: scripts/linux/emit-vector-piggy.py [hunk 1]
```diff
@@ -0,0 +1,67 @@
...
+#
+# * You add a bitmap
+# * You add a resolution
+
+import os
+import re
+
+assets_path = "assets/original-vector/exported"
+
+source_file = open("src/common/ScalablePiggy.S", "w")
```
Comment (baconpaul) [inline scripts/linux/emit-vector-piggy.py]: src/linux/ surely?

File: src/common/gui/CScalableBitmap.cpp [hunk 2]
```diff
@@ -16,13 +34,8 @@ void CScalableBitmap::setPhysicalZoomFactor(int zoomFactor)
     currentPhysicalZoomFactor = zoomFactor;
 }
 
-#if __linux__
 CScalableBitmap::CScalableBitmap(CResourceDescription desc)
-    : CBitmap(originalPlatformBitmapForID(desc.u.id))
-#else
-CScalableBitmap::CScalableBitmap(CResourceDescription desc) 
     : CBitmap(desc)
-#endif
```
Comment (baconpaul) [inline src/common/gui/CScalableBitmap.cpp]: I'm surprised this diff works; On linux you will end up with a base class for this instance with size 0 won't you? So then clients of CScalableBitmap who ask ->getWidth() will get a 0 even though it has drawable size.

I know on Mac having a non-valid superclass constructor caused mis-draws and crashes, which is why I kept this here there.

**Comment by baconpaul** [comment]
Looking good. I tossed in a few review comments.

File: premake5.lua [hunk 3]
```diff
@@ -87,6 +87,11 @@ elseif (os.istarget("linux")) then
 
 	platforms { "x64" }
 
+    files
+    {
+        "src/common/ScalablePiggy.S"
```
[the remark this answers is not in the document]
Comment (jsakkine) [reply premake5.lua]: I don't know. There is no real logic with these right now. Like scalableresource.h is also in common. I found best to put them into same location (.h and .S).

File: src/common/gui/CScalableBitmap.cpp (hunk 2 above)
Comment (baconpaul) [reply src/common/gui/CScalableBitmap.cpp]: Huh cool but from your screenshot it does work. OK this one is fine!

File: premake5.lua (hunk 3 above)
[the remark this answers is not in the document]
Comment (baconpaul) [reply premake5.lua]: but the scalable .rc file is not in src/common. Isn't this more like the .rc file?

(I could move scalable resource.h into windows also and conditionally include it in resource.h).

[the remark this answers is not in the document]
Comment (jsakkine) [reply premake5.lua]: I think it should be done, as the header is neither common by any means. I'll do the same for this.

**Comment by baconpaul** [comment]
Ha this really is cool!

Like I said just minor comments about putting files in the right place. Happy to sweep this if we just mod build_linux.sh (and really, i think we should move the piggy files; the more I think about it I was wrong to put scalableresource.h in src/common and I'll fix/move that tomorrow too).

**Comment by jsakkine** [comment]
I'll check that macOS still working before dealing with your comments. They are sane proposals, no worries.

**Comment by jsakkine** [comment]
Yea, macOS works just fine!

**Comment by baconpaul** [comment]
Oh mzcwill be fine Yu restoredit to this morning state don't worry

**Comment by jsakkine** [comment]
Is it good now?

File: src/common/gui/CScalableBitmap.cpp [hunk 4]
```diff
@@ -1,13 +1,31 @@
...
 #include "resource.h"
 
 using namespace VSTGUI;
 
+#if __linux__
+static const struct MemoryBitmap *findMemoryBitmap(const std::string& filename)
+{
+    for (const struct MemoryBitmap *bmp = &memoryBitmapList[0];
+         bmp->name != NULL; bmp++)
+        if (!std::strncmp(filename.c_str(), bmp->name, strlen(bmp->name)))
```
[the remark this answers is not in the document]
Comment (jsakkine) [reply src/common/gui/CScalableBitmap.cpp]: Thanks, fixed.

**Pull request merged**